### PR TITLE
Plugins: Refactor plugin download/installation 

### PR DESCRIPTION
**Pull request by wbrowne**
**What this PR does / why we need it**:

-  The existing `Installer` component is now broken up into two new interfaces - `repo.Service` and `storage.Manager`.  The `repo` package encapsulates communication with a plugin repository (grafana.com API for example) and the `storage` package provides an abstraction for how plugin artefacts are stored and managed (on disk, object storage, etc.)
- Public grafana plugin "SDK" now accepts "compatibility options" when installing plugins. Previously, we were hardcoding this info and using Grafana server specific config, which hurts portability (now we're closer to not depending on any specific Grafana config within the `plugins` package)
- Fixes a bug where if you request to install a plugin (without specifying a version), and that specific plugin version is already installed, then abort the install (before it would just re-install it)

**Special notes for your reviewer**:
- Manually testing out plugin install and uninstall (via the HTTP API/plugin catalog and `grafana-cli`) would be helpful 🙇

### CODE REVIEW

File: pkg/api/plugins.go [hunk 1]
```diff
@@ -368,21 +370,25 @@ func (hs *HTTPServer) InstallPlugin(c *models.ReqContext) response.Response {
 	}
 	pluginID := web.Params(c.Req)[":pluginId"]
 
-	err := hs.pluginManager.Add(c.Req.Context(), pluginID, dto.Version)
+	err := hs.pluginManager.Add(c.Req.Context(), pluginID, dto.Version, plugins.CompatOpts{
+		GrafanaVersion: hs.Cfg.BuildVersion,
```
Comment (marefr) [inline pkg/api/plugins.go:374]: Isn't this always the same value, both here and in cli (services.GrafanaVersion)? I.e. the version injected when building Grafana? I guess this related to your description (below), but I don't understand what you mean, sorry.

> Public grafana plugin "SDK" now accepts "compatibility options" when installing plugins. Previously, we were hardcoding this info and using Grafana server specific config, which hurts portability (now we're closer to not depending on any specific Grafana config within the plugins package)

File: pkg/plugins/storage/fs.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+		// Check for ZipSlip. More Info: http://bit.ly/2MsjAWE
+		if filepath.IsAbs(zf.Name) ||
+			!strings.HasPrefix(fullPath, filepath.Clean(fs.pluginsDir)+string(os.PathSeparator)) ||
+			strings.HasPrefix(zf.Name, ".."+string(os.PathSeparator)) {
+			return "", fmt.Errorf(
+				"archive member %q tries to write outside of plugin directory: %q, this can be a security risk",
+				zf.Name, fs.pluginsDir)
+		}
+
+		dstPath := filepath.Clean(filepath.Join(fs.pluginsDir, removeGitBuildFromName(zf.Name, pluginID)))
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/plugins/storage/fs.go:130]: Probably need to double check this one

**Review by marefr — APPROVED** [review]
Looks good. Added a couple comments

File: pkg/api/plugins.go (hunk 1 above)
Comment (wbrowne) [reply pkg/api/plugins.go:374]: The intention here is to make the APIs more flexible depending on the client. Core grafana will always have the same version, but in other cases, maybe/maybe not 😉

Comment (marefr) [reply pkg/api/plugins.go:374]: Got it :+1:

File: pkg/plugins/storage/fs.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (wbrowne) [reply pkg/plugins/storage/fs.go]: Think we're good here based on your previous work https://github.com/grafana/grafana/pull/50537/files#diff-5c0fc73f5780d4f458dd3d0ceaa361e2fbd2134f81976ecc940c926eba07b97eR611. Have added a codeql ignore

**Pull request merged**